File: app/src/main/java/org/tvheadend/tvhclient/data/source/ChannelData.java
package org.tvheadend.tvhclient.data.source;

import android.os.AsyncTask;

import org.tvheadend.tvhclient.data.db.AppRoomDatabase;
import org.tvheadend.tvhclient.data.entity.Channel;
import org.tvheadend.tvhclient.data.entity.EpgChannel;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutionException;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.LiveData;
import timber.log.Timber;

public class ChannelData implements DataSourceInterface<Channel> {

    private final AppRoomDatabase db;

    public ChannelData(AppRoomDatabase database) {
        this.db = database;
    }

    @Override
    public void addItem(Channel item) {
        AsyncTask.execute(() -> db.getChannelDao().insert(item));
    }

    public void addItems(@NonNull List<Channel> items) {
        AsyncTask.execute(() -> db.getChannelDao().insert(new ArrayList<>(items)));
    }

    @Override
    public void updateItem(Channel item) {
        AsyncTask.execute(() -> db.getChannelDao().update(item));
    }

    @Override
    public void removeItem(Channel item) {
        AsyncTask.execute(() -> db.getChannelDao().delete(item));
    }

    @Override
    public LiveData<Integer> getLiveDataItemCount() {
        return db.getChannelDao().getItemCount();
    }

    @Override
    public LiveData<List<Channel>> getLiveDataItems() {
        return null;
    }

    @Override
    public LiveData<Channel> getLiveDataItemById(Object id) {
        return null;
    }

    @Override
    @Nullable
    public Channel getItemById(Object id) {
        try {
            return new ChannelByIdTask(db, (int) id).execute().get();
        } catch (InterruptedException e) {
            Timber.d("Loading channel by id task got interrupted", e);
        } catch (ExecutionException e) {
            Timber.d("Loading channel by id task aborted", e);
        }
        return null;
    }

    @Override
    @NonNull
    public List<Channel> getItems() {
        List<Channel> channels = new ArrayList<>();
        try {
            channels.addAll(new ChannelListTask(db).execute().get());
        } catch (InterruptedException e) {
            Timber.d("Loading all channels task got interrupted", e);
        } catch (ExecutionException e) {
            Timber.d("Loading all channels task aborted", e);
        }
        return channels;
    }

    public int getItemCount() {
        try {
            return new ChannelCountTask(db).execute().get();
        } catch (InterruptedException e) {
            Timber.d("Loading channel count task got interrupted", e);
        } catch (ExecutionException e) {
            Timber.d("Loading channel count task aborted", e);
        }
        return 0;
    }

    @Nullable
    public Channel getItemByIdWithPrograms(int id, long selectedTime) {
        try {
            return new ChannelByIdTask(db, id, selectedTime).execute().get();
        } catch (InterruptedException e) {
            Timber.d("Loading channel by id task got interrupted", e);
        } catch (ExecutionException e) {
            Timber.d("Loading channel by id task aborted", e);
        }
        return null;
    }

    public LiveData<List<EpgChannel>> getAllEpgChannels(int channelSortOrder, @NonNull List<Integer> tagIds) {
        Timber.d("Loading epg channels with sort order " + channelSortOrder + " and " + tagIds.size() + " tags");
        if (tagIds.size() == 0) {
            return db.getChannelDao().loadAllEpgChannels(channelSortOrder);
        } else {
            return db.getChannelDao().loadAllEpgChannelsByTag(channelSortOrder, tagIds);
        }
    }

    public LiveData<List<Channel>> getAllChannelsByTime(long selectedTime, int channelSortOrder, @NonNull List<Integer> tagIds) {
        Timber.d("Loading channels from time " + selectedTime + " with sort order " + channelSortOrder + " and " + tagIds.size() + " tags");
        if (tagIds.size() == 0) {
            return db.getChannelDao().loadAllChannelsByTime(selectedTime, channelSortOrder);
        } else {
            return db.getChannelDao().loadAllChannelsByTimeAndTag(selectedTime, channelSortOrder, tagIds);
        }
    }

    private static class ChannelByIdTask extends AsyncTask<Void, Void, Channel> {
        private final AppRoomDatabase db;
        private final int id;
        private final long selectedTime;

        ChannelByIdTask(AppRoomDatabase db, int id) {
            this.db = db;
            this.id = id;
            this.selectedTime = 0;
        }

        ChannelByIdTask(AppRoomDatabase db, int id, long selectedTime) {
            this.db = db;
            this.id = id;
            this.selectedTime = selectedTime;
        }

        @Override
        protected Channel doInBackground(Void... voids) {
            if (selectedTime > 0) {
                return db.getChannelDao().loadChannelByIdWithProgramsSync(id, selectedTime);
            } else {
                return db.getChannelDao().loadChannelByIdSync(id);
            }
        }
    }

    private static class ChannelListTask extends AsyncTask<Void, Void, List<Channel>> {
        private final AppRoomDatabase db;

        ChannelListTask(AppRoomDatabase db) {
            this.db = db;
        }

        @Override
        protected List<Channel> doInBackground(Void... voids) {
            return db.getChannelDao().loadAllChannelsSync(0);
        }
    }

    private static class ChannelCountTask extends AsyncTask<Void, Void, Integer> {
        private final AppRoomDatabase db;

        ChannelCountTask(AppRoomDatabase db) {
            this.db = db;
        }

        @Override
        protected Integer doInBackground(Void... voids) {
            return db.getChannelDao().getItemCountSync();
        }
    }
}

File: app/src/main/java/org/tvheadend/tvhclient/domain/repository/data_source/ChannelData.kt
package org.tvheadend.tvhclient.domain.repository.data_source

import android.os.AsyncTask
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import org.tvheadend.tvhclient.data.db.AppRoomDatabase
import org.tvheadend.tvhclient.domain.entity.Channel
import org.tvheadend.tvhclient.domain.entity.EpgChannel
import timber.log.Timber
import java.util.*
import java.util.concurrent.ExecutionException

class ChannelData(private val db: AppRoomDatabase) : DataSourceInterface<Channel> {

    val itemCount: Int
        get() {
            try {
                return ChannelCountTask(db).execute().get()
            } catch (e: InterruptedException) {
                Timber.e(e, "Loading channel count task got interrupted")
            } catch (e: ExecutionException) {
                Timber.e(e, "Loading channel count task aborted")
            }

            return 0
        }

    override fun addItem(item: Channel) {
        AsyncTask.execute { db.channelDao.insert(item) }
    }

    fun addItems(items: List<Channel>) {
        AsyncTask.execute { db.channelDao.insert(ArrayList(items)) }
    }

    override fun updateItem(item: Channel) {
        AsyncTask.execute { db.channelDao.update(item) }
    }

    override fun removeItem(item: Channel) {
        AsyncTask.execute { db.channelDao.delete(item) }
    }

    fun removeItemById(id: Int) {
        AsyncTask.execute { db.channelDao.deleteById(id) }
    }

    override fun getLiveDataItemCount(): LiveData<Int> {
        return db.channelDao.itemCount
    }

    override fun getLiveDataItems(): LiveData<List<Channel>> {
        return MutableLiveData()
    }

    override fun getLiveDataItemById(id: Any): LiveData<Channel> {
        return MutableLiveData()
    }

    override fun getItemById(id: Any): Channel? {
        try {
            return ChannelByIdTask(db, id as Int).execute().get()
        } catch (e: InterruptedException) {
            Timber.e(e, "Loading channel by id task got interrupted")
        } catch (e: ExecutionException) {
            Timber.e(e, "Loading channel by id task aborted")
        }

        return null
    }

    fun getChannels(sortOrder: Int = 0): List<Channel> {
        val channels = ArrayList<Channel>()
        try {
            channels.addAll(ChannelListTask(db, sortOrder).execute().get())
        } catch (e: InterruptedException) {
            Timber.e(e, "Loading all channels task got interrupted")
        } catch (e: ExecutionException) {
            Timber.e(e, "Loading all channels task aborted")
        }

        return channels
    }

    override fun getItems(): List<Channel> {
        return getChannels()
    }

    fun getItemByIdWithPrograms(id: Int, selectedTime: Long): Channel? {
        try {
            return ChannelByIdTask(db, id, selectedTime).execute().get()
        } catch (e: InterruptedException) {
            Timber.e(e, "Loading channel by id task got interrupted")
        } catch (e: ExecutionException) {
            Timber.e(e, "Loading channel by id task aborted")
        }

        return null
    }

    fun getAllEpgChannels(channelSortOrder: Int, tagIds: List<Int>): LiveData<List<EpgChannel>> {
        Timber.d("Loading epg channels with sort order $channelSortOrder and ${tagIds.size} tags")
        return if (tagIds.isEmpty()) {
            db.channelDao.loadAllEpgChannels(channelSortOrder)
        } else {
            db.channelDao.loadAllEpgChannelsByTag(channelSortOrder, tagIds)
        }
    }

    fun getAllChannelsByTime(selectedTime: Long, channelSortOrder: Int, tagIds: List<Int>): LiveData<List<Channel>> {
        Timber.d("Loading channels from time $selectedTime with sort order $channelSortOrder and ${tagIds.size} tags")
        return if (tagIds.isEmpty()) {
            db.channelDao.loadAllChannelsByTime(selectedTime, channelSortOrder)
        } else {
            db.channelDao.loadAllChannelsByTimeAndTag(selectedTime, channelSortOrder, tagIds)
        }
    }

    private class ChannelByIdTask : AsyncTask<Void, Void, Channel> {
        private val db: AppRoomDatabase
        private val id: Int
        private val selectedTime: Long

        internal constructor(db: AppRoomDatabase, id: Int) {
            this.db = db
            this.id = id
            this.selectedTime = 0
        }

        internal constructor(db: AppRoomDatabase, id: Int, selectedTime: Long) {
            this.db = db
            this.id = id
            this.selectedTime = selectedTime
        }

        override fun doInBackground(vararg voids: Void): Channel {
            return if (selectedTime > 0) {
                db.channelDao.loadChannelByIdWithProgramsSync(id, selectedTime)
            } else {
                db.channelDao.loadChannelByIdSync(id)
            }
        }
    }

    private class ChannelListTask internal constructor(private val db: AppRoomDatabase, private val sortOrder: Int) : AsyncTask<Void, Void, List<Channel>>() {

        override fun doInBackground(vararg voids: Void): List<Channel> {
            return db.channelDao.loadAllChannelsSync(sortOrder)
        }
    }

    private class ChannelCountTask internal constructor(private val db: AppRoomDatabase) : AsyncTask<Void, Void, Int>() {

        override fun doInBackground(vararg voids: Void): Int {
            return db.channelDao.itemCountSync
        }
    }
}
